File: hashtable/525.go
func findMaxLength(nums []int) int {
    var ret, count int
    m := make(map[int]int)
    m[0] = -1
    for i := 0; i < len(nums); i++ {
        if nums[i] == 0 {
            count--
        }else {
            count++
        }
        if _, ok := m[count]; ok {
            if i - m[count] > ret {
                ret = i - m[count];
            }
        } else {
            m[count] = i
        }
    }
    return ret
}


func findMaxLength(nums []int) int {
    var count, ret int
    m := make(map[int]int)
    m[0] = -1
    for i, v := range nums {
        if v == 0 {
            count--
        }else {
            count++
        }
        if index, ok := m[count]; ok {
            if i - index > ret {
                ret = i - index
            }
        }else {
            m[count] = i
        }
    }
    return ret    
}

File: hashtable/525.cc
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        std::unordered_map<int, int> pos;
        pos[0] = -1;
        int count = 0, ret = 0;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] == 0) {
                ++count;
            } else {
                --count;
            }
            auto iter = pos.find(count);
            if (iter != pos.end()) {
                ret = std::max(ret, i - iter->second);    
            } else {
                pos[count] = i;
            }
        }
        return ret;
    }
};
